File: generated_code_examples/c/classification/xgboost.c
#include <string.h>
void softmax(double *x, int size, double *result) {
    double max = x[0];
    for (int i = 1; i < size; ++i) {
        if (x[i] > max)
            max = x[i];
    }
    double sum = 0.0;
    for (int i = 0; i < size; ++i) {
        result[i] = exp(x[i] - max);
        sum += result[i];
    }
    for (int i = 0; i < size; ++i)
        result[i] /= sum;
}
void score(double * input, double * output) {
    double var0[3];
    double var1;
    if (input[2] >= 2.45) {
        var1 = -0.21995015;
    } else {
        var1 = 0.4302439;
    }
    double var2;
    if (input[2] >= 2.45) {
        var2 = -0.19691855;
    } else {
        var2 = 0.29493433;
    }
    double var3;
    if (input[2] >= 2.45) {
        if (input[3] >= 1.75) {
            var3 = -0.20051816;
        } else {
            var3 = 0.36912444;
        }
    } else {
        var3 = -0.21512198;
    }
    double var4;
    if (input[2] >= 2.45) {
        if (input[2] >= 4.8500004) {
            var4 = -0.14888482;
        } else {
            var4 = 0.2796613;
        }
    } else {
        var4 = -0.19143805;
    }
    double var5;
    if (input[3] >= 1.6500001) {
        var5 = 0.40298507;
    } else {
        if (input[2] >= 4.95) {
            var5 = 0.21724138;
        } else {
            var5 = -0.21974029;
        }
    }
    double var6;
    if (input[2] >= 4.75) {
        if (input[3] >= 1.75) {
            var6 = 0.28692952;
        } else {
            var6 = 0.06272897;
        }
    } else {
        if (input[3] >= 1.55) {
            var6 = 0.009899145;
        } else {
            var6 = -0.19659369;
        }
    }
    softmax((double[]){0.5 + (var1 + var2), 0.5 + (var3 + var4), 0.5 + (var5 + var6)}, 3, var0);
    memcpy(output, var0, 3 * sizeof(double));
}

File: generated_code_examples/fortran/classification/xgboost.f90
module Model
    implicit none
contains
    function score(input)
        implicit none
        double precision, dimension(:) :: input
        double precision :: var0
        double precision :: var1
        double precision :: var2
        double precision :: var3
        double precision :: var4
        double precision :: var5
        double precision, dimension(3) :: score
        if (input(3) >= 2.45d0) then
            var0 = -0.21995015d0
        else
            var0 = 0.4302439d0
        end if
        if (input(3) >= 2.45d0) then
            var1 = -0.19691855d0
        else
            var1 = 0.29493433d0
        end if
        if (input(3) >= 2.45d0) then
            if (input(4) >= 1.75d0) then
                var2 = -0.20051816d0
            else
                var2 = 0.36912444d0
            end if
        else
            var2 = -0.21512198d0
        end if
        if (input(3) >= 2.45d0) then
            if (input(3) >= 4.8500004d0) then
                var3 = -0.14888482d0
            else
                var3 = 0.2796613d0
            end if
        else
            var3 = -0.19143805d0
        end if
        if (input(4) >= 1.6500001d0) then
            var4 = 0.40298507d0
        else
            if (input(3) >= 4.95d0) then
                var4 = 0.21724138d0
            else
                var4 = -0.21974029d0
            end if
        end if
        if (input(3) >= 4.75d0) then
            if (input(4) >= 1.75d0) then
                var5 = 0.28692952d0
            else
                var5 = 0.06272897d0
            end if
        else
            if (input(4) >= 1.55d0) then
                var5 = 0.009899145d0
            else
                var5 = -0.19659369d0
            end if
        end if
        score(:) = SOFTMAX((/ 0.5d0 + var0 + var1, 0.5d0 + var2 + var3, 0.5d0 + var4 + var5 /))
        return
    end function score
    function softmax(x) result(res)
        implicit none
        double precision, dimension(:), intent(in) :: x
        double precision, dimension(size(x)) :: res
        double precision :: max_val, sum_val
        integer :: i

        ! Find maximum value in x
        max_val = x(1)
        do i = 2, size(x)
            if (x(i) > max_val) then
                max_val = x(i)
            end if
        end do

        ! Compute softmax values
        sum_val = 0.0d0
        do i = 1, size(x)
            res(i) = exp(x(i) - max_val)
            sum_val = sum_val + res(i)
        end do
        res = res / sum_val

    end function softmax
end module Model
